synapse/handlers/federation.py

```diff
@@ -1705,8 +1719,31 @@ async def _sync_partial_state_room_wrapper() -> None:
                     room_id=room_id,
                 )
             finally:
+                # Check whether the room is still partial stated, while we still claim
+                # to be the active sync. Usually, the partial state flag will be gone,
+                # unless we left and rejoined the room, or the sync failed.
+                is_still_partial_state_room = await self.store.is_partial_state_room(
+                    room_id
+                )
                 self._active_partial_state_syncs.remove(room_id)
 
+                # Check if we need to restart the sync.
+                if room_id in self._partial_state_syncs_to_restart:
+                    (
+                        restart_initial_destination,
+                        restart_other_destinations,
+                    ) = self._partial_state_syncs_to_restart[room_id]
+
+                    # Clear the restart flag.
+                    self._partial_state_syncs_to_restart.pop(room_id, None)
+
+                    if is_still_partial_state_room:
+                        self._start_partial_state_room_sync(
+                            initial_destination=restart_initial_destination,
+                            other_destinations=restart_other_destinations,
+                            room_id=room_id,
+                        )
+
```

Comment:
Suppose that the needs-restart flag is set, but `is_still_partial_state_room` is False. In this situation we clear the flag but don't restart the sync. That seems surprising... but maybe it's correct? What's going on here.

(Maybe it's not a `needs-restart` flag but a `might-need-restart` flag?)

Reply:
Another way of looking at it might be:
if the partial state flag isn't set, there's no (remaining) work for the sync to do, so there's no point in restarting it